controller/metrics/clustersecretupdater.go

```diff
@@ -0,0 +1,37 @@
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/argoproj/argo-cd/util/db"
+)
+
+type clusterSecretUpdater struct {
+	infoSource HasClustersInfo
+	lock       sync.Mutex
+	db         db.ArgoDB
+}
+
+func (c clusterSecretUpdater) Run(ctx context.Context) {
+	tick := time.Tick(metricsCollectionInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			break
+		case <-tick:
+
+			for _, info := range c.infoSource.GetClustersInfo() {
+				cluster, err := c.db.GetCluster(context.Background(), info.Server)
+				if err != nil {
+					continue
+				}
+				cluster.ServerVersion = info.K8SVersion
+				cluster.LastCacheSyncTime = info.LastCacheSyncTime.String()
```

Comment:
The controller might be unable to connect to cluster. In this case the status should be failed with the appropriate error message. The connection error is not available in ClusterInfo data structure but it is easy to add it:

https://github.com/argoproj/argo-cd/blob/8ad928330f26b3eea523626d306306e4929aaa7a/controller/cache/cluster.go#L675

Lets store both ConnectionState and Message in cluster secret annotations